finesse/hardware/plugins/spectrometer/ftsw500_interface.py

```diff
@@ -70,135 +60,141 @@ def __init__(self) -> None:
-                self._status = new_status
-                self.send_status_message(new_status)
+        retval = self._make_request("getFTSW500State")
+
+        try:
+            status_num = int(retval)
+
+            if status_num == -1:
+                # Try again later
+                return None
+            if 0 <= status_num <= 3:
+                return SpectrometerStatus(status_num)
+        except ValueError:
+            pass
 
+        raise FTSW500Error(f"Invalid value received for status: {retval}")
+
+    def _update_status(self) -> None:
+        """Update the current status.
+
+        The status polling timer is restarted to ensure that another status update will
+        happen soon. If the status is intermediate (-1), then it is ignored and we just
+        wait until the status is next requested.
+        """
+        new_status = self._get_status()
+
+        # If the status is not intermediate and has changed since we last checked
+        if new_status and new_status != self._status:
+            self._status = new_status
+            self.send_status_message(new_status)
+
+        # Request another status update soon
         self._status_timer.start()
 
-        self._regurgitate_nonmodal_dialog_message()
-        self._regurgitate_modal_dialog_message()
+    def _make_request(self, command: str) -> str:
+        """Request that FTSW500 run the specified command.
+
+        Args:
+            command: Name of command to run
+        """
+        self._socket.sendall(f"{command}\n".encode())
+
+        # Read a single message, which should be terminated with a newline. As the
+        # protocol is simple we don't have to worry about what happens if more than or
+        # less than one line of text is received.
+        response = self._socket.recv(1024).decode()
+        if not response:
+            raise FTSW500Error("Connection terminated unexpectedly")
 
-    def _make_request(self, command) -> None:
-        """Make a request."""
-        self._requester.sendall(command)
-        self._on_reply_received(self._requester.recv(1024))
+        if not response.endswith("\n"):
+            raise FTSW500Error("Response not terminated with newline")
 
-    def _request_status(self) -> None:
-        """Request the current status from FTSW500."""
-        self.request_command(b"getFTSW500State\n")
+        return _parse_response(response[:-1])
 
-    def request_command(self, command: bytes) -> None:
+    def request_command(self, command: str) -> None:
         """Request that FTSW500 run the specified command.
 
+        The status is requested after sending the command so that we are immediately
+        notified if it has changed in response to the command (e.g. recording has
+        started). Contents of dialogs in FTSW500 are saved to the program log.
+
         Args:
             command: Name of command to run
         """
-        self._make_request(command)
+
+        def internal():
```

Comment:
Why making this an internal function? That makes it directly untestable.

Reply:
Good point! I'll fix that up